packages/babel-plugin-transform-react-constant-elements/src/index.js

```diff
@@ -83,30 +160,70 @@ export default declare((api, options) => {
-        HOISTED.add(path.node);
+        HOISTED.set(path.node, path.scope);
 
-        const state = { isImmutable: true };
+        const name = path.node.openingElement.name;
 
         // This transform takes the option `allowMutablePropsOnTags`, which is an array
         // of JSX tags to allow mutable props (such as objects, functions) on. Use sparingly
         // and only on tags you know will never modify their own props.
+        let mutablePropsAllowed = false;
         if (allowMutablePropsOnTags != null) {
           // Get the element's name. If it's a member expression, we use the last part of the path.
           // So the option ["FormattedMessage"] would match "Intl.FormattedMessage".
-          let namePath = path.get("openingElement.name");
-          while (namePath.isJSXMemberExpression()) {
-            namePath = namePath.get("property");
+          let lastSegment = name;
+          while (t.isJSXMemberExpression(lastSegment)) {
+            lastSegment = lastSegment.property;
           }
 
-          const elementName = namePath.node.name;
-          state.mutablePropsAllowed =
-            allowMutablePropsOnTags.indexOf(elementName) > -1;
+          const elementName = lastSegment.name;
+          mutablePropsAllowed = allowMutablePropsOnTags.includes(elementName);
+        }
+
+        // In order to avoid hoisting unnecessarily, we need to know which is
+        // the scope containing the current JSX element. If a parent of the
+        // current element has already been hoisted, we can consider its target
+        // scope as the base scope for the current element.
+        let jsxScope;
+        let current = path;
+        while (!jsxScope && current.parentPath.isJSXElement()) {
+          current = current.parentPath;
+          jsxScope = HOISTED.get(current.node);
         }
+        jsxScope ??= path.scope;
+
+        const state = {
+          isImmutable: true,
+          mutablePropsAllowed,
+          jsxScope,
+          targetScope: null,
+        };
+
+        // Traverse all props passed to this element for immutability,
+        // and compute the target hoisting scope
+        path.traverse(analyzer, state);
 
-        // Traverse all props passed to this element for immutability.
-        path.traverse(immutabilityVisitor, state);
+        if (!state.isImmutable) return;
+
+        // If we didn't find any hoisting constraint, we can hoist the current
+        // helement to the program scope.
+        const targetScope = state.targetScope ?? path.scope.getProgramParent();
+        HOISTED.set(path.node, targetScope);
+
+        // Only hoist if it would give us an advantage.
+        if (targetScope === jsxScope) return;
+
+        const id = path.scope.generateUidBasedOnNode(name);
+        targetScope.push({ id: t.identifier(id) });
+
+        let replacement = template.expression.ast`
+          ${t.identifier(id)} || (${t.identifier(id)} = ${path.node})
```

Comment:
nit: we can output `${t.identifier(id)} ||= ${path.node}` if the `target` has native logical assignment support, or do it by default in Babel 8.